<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Matthew Hung">
<meta name="dcterms.date" content="2024-11-08">

<title>Single-Cell Workbook - 5&nbsp; Cell Demultiplexing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../sections/1_processing/1.4_qualitycontrol.html" rel="next">
<link href="../../sections/1_processing/1.2_seurat.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../sections/1_processing/1.1_cellranger.html"><strong>Chapter 1 : Data Processing</strong></a></li><li class="breadcrumb-item"><a href="../../sections/1_processing/1.3_demux.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Cell Demultiplexing</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-Cell Workbook</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/mshung229/scworkbook/">
            Report A Bug
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/mshung229">
            Follow Me
            </a>
          </li>
      </ul>
    </div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Preface</b></span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text"><strong>Chapter 0 : Introduction</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/0_introduction/0.1_installation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Installation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/0_introduction/0.2_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Description</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text"><strong>Chapter 1 : Data Processing</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.1_cellranger.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">CellRanger (Multi)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.2_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Seurat Object</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.3_demux.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Cell Demultiplexing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.4_qualitycontrol.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.5_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Unsupervised Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.6_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.7_doublets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Doublet Detection</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#motivation" id="toc-motivation" class="nav-link active" data-scroll-target="#motivation">Motivation</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup"><span class="header-section-number">5.1</span> Setup</a></li>
  <li><a href="#sec-calculate_gene_fractions" id="toc-sec-calculate_gene_fractions" class="nav-link" data-scroll-target="#sec-calculate_gene_fractions"><span class="header-section-number">5.2</span> Calculate Gene Fractions</a></li>
  <li><a href="#inpect-hto-counts" id="toc-inpect-hto-counts" class="nav-link" data-scroll-target="#inpect-hto-counts"><span class="header-section-number">5.3</span> Inpect HTO Counts</a></li>
  <li><a href="#sec-htodemux" id="toc-sec-htodemux" class="nav-link" data-scroll-target="#sec-htodemux"><span class="header-section-number">5.4</span> Run HTODemux</a></li>
  <li><a href="#sec-multiseqdemux" id="toc-sec-multiseqdemux" class="nav-link" data-scroll-target="#sec-multiseqdemux"><span class="header-section-number">5.5</span> Run MULTIseqDemux</a></li>
  <li><a href="#sec-htosignal" id="toc-sec-htosignal" class="nav-link" data-scroll-target="#sec-htosignal"><span class="header-section-number">5.6</span> Separate HTO Signals</a></li>
  <li><a href="#evaluation" id="toc-evaluation" class="nav-link" data-scroll-target="#evaluation"><span class="header-section-number">5.7</span> Evaluation</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../sections/1_processing/1.1_cellranger.html"><strong>Chapter 1 : Data Processing</strong></a></li><li class="breadcrumb-item"><a href="../../sections/1_processing/1.3_demux.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Cell Demultiplexing</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Cell Demultiplexing</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Matthew Hung </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 8, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="motivation" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="motivation">Motivation</h2>
<p>For this dataset, each sample is comprised of cells from multiple patients hashtagged with a unique oligo sequence. To demultiplex which patient the cells were retrieved from, a demultiplexing algorithm must be used to quantify the oligo sequences for each cell and assign each cell to a patient. The most common approaches to date is using the HTOdemux <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1603-1">[Stoeckius et al.; 2018]</a> or the MULTIseqDemux <a href="https://www.nature.com/articles/s41592-019-0433-8">[McGinnis et al.; 2019]</a> algorithms. Besides assigning each cell back to each patient, these algorithms are able to identify homolytic/heterolytic “Doublet” and “Negative” cells based on high/multiple and low oligo counts respectively. These poor quality droplets are a common issue with droplet-based sequencing when there are multiple or no cells found in each droplet, and removal of these droplets could prevent false-positive findings.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aim
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Run HTODemux, MULTIseqDemux pipelines to demultiplex cells.</li>
<li>Compare their strengths and efficiency.</li>
</ul>
</div>
</div>
</section>
<section id="setup" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="setup"><span class="header-section-number">5.1</span> Setup</h2>
<div id="cell-3" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set up environment</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>({</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scUnify)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"/nemo/lab/caladod/working/Matthew/project/matthew/MH_GSE247917"</span>)})</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># load seurat objects</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>obj.list <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"seurat/1_processing/1.2_GSE247917_raw.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“replacing previous import ‘cowplot::get_legend’ by ‘ggpubr::get_legend’ when loading ‘scUnify’”
Warning message:
“replacing previous import ‘cowplot::align_plots’ by ‘patchwork::align_plots’ when loading ‘scUnify’”
Warning message:
“replacing previous import ‘biomaRt::select’ by ‘rstatix::select’ when loading ‘scUnify’”
Warning message:
“replacing previous import ‘scales::viridis_pal’ by ‘viridis::viridis_pal’ when loading ‘scUnify’”</code></pre>
</div>
</div>
<div id="cell-4" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect object and make sure active assay is RNA</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>obj.list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>$CV10
An object of class Seurat 
23862 features across 19484 samples within 3 assays 
Active assay: RNA (23797 features, 0 variable features)
 1 layer present: counts
 2 other assays present: HTO, ADT

$CV12
An object of class Seurat 
16451 features across 17366 samples within 3 assays 
Active assay: RNA (16386 features, 0 variable features)
 1 layer present: counts
 2 other assays present: HTO, ADT</code></pre>
</div>
</div>
</section>
<section id="sec-calculate_gene_fractions" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="sec-calculate_gene_fractions"><span class="header-section-number">5.2</span> Calculate Gene Fractions</h2>
<p>Before demultiplexing, we need to first calculate the fraction of mitochondrial, ribosomal, haemoglobin, T/B-cell recepter (TCR/BCR) and MHC reads to identify poor quality cells, which can be informative of accurate demultiplexing. The function below will perform the calculations of key parameters that identifies poor quality cells.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Please navigate to the next chapter &lt;Quality Control&gt; for a detailed explanation of the wrapper function, outputs, and interpretations.</p>
</div>
</div>
<div id="cell-6" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## calculate gene fractions (mt, rb, hb, tcr, bcr, mhc)</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.list)){</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    obj.list[[x]] <span class="ot">&lt;-</span> <span class="fu">calculate_fractions</span>(obj.list[[x]])}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="inpect-hto-counts" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="inpect-hto-counts"><span class="header-section-number">5.3</span> Inpect HTO Counts</h2>
<p>Before demultiplexing HTO library, we will inspect and remove any redundant/low expressing hashtags. For each sample, below we will quantify the expression levels for for each hashtag in each sample.</p>
<blockquote class="blockquote">
<p>Quantification show that all hashtags are abundantly expressed across the samples and can therefore proceed to demultiplexing.</p>
</blockquote>
<div id="cell-8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate how many cells expresses each tag</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>metalist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.list)){</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    metalist[[x]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample =</span> <span class="fu">names</span>(obj.list)[x], <span class="at">count =</span> <span class="fu">apply</span>(obj.list[[x]][[<span class="st">"HTO"</span>]]<span class="sc">$</span>counts, <span class="dv">1</span>, <span class="cf">function</span>(row) <span class="fu">sum</span>(row <span class="sc">!=</span> <span class="dv">0</span>)))}</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>metalist <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(metalist)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>metalist</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A data.frame: 8 × 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">sample</th>
<th data-quarto-table-cell-role="th" scope="col">count</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;int&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">CV-011-d120-booster</td>
<td>CV10</td>
<td>16404</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">CV-001-d28-booster</td>
<td>CV10</td>
<td>16189</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">CV-001-d7-booster</td>
<td>CV10</td>
<td>19476</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">CV-001-d0-booster</td>
<td>CV10</td>
<td>17218</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">CV-053-d28-vax</td>
<td>CV12</td>
<td>9232</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">CV-053-d21-vax</td>
<td>CV12</td>
<td>12223</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">CV-053-d7-vax</td>
<td>CV12</td>
<td>10337</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">CV-053-d0-vax</td>
<td>CV12</td>
<td>11173</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="sec-htodemux" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="sec-htodemux"><span class="header-section-number">5.4</span> Run HTODemux</h2>
<p>First we will run the the wrapper function below to demultiplex the cells with the HTODemux algorithm in default settings.</p>
<div id="cell-10" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># if necessary, add one pseudocount to each oligo to run HTODemux without error</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.list)){</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    obj.list[[x]] <span class="ot">&lt;-</span> <span class="fu">run_htodemux</span>(obj.list[[x]], <span class="at">add.one =</span> F)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts

Normalizing across cells

As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.
This message is displayed once per session.
Cutoff for CV-011-d120-booster : 57 reads

Cutoff for CV-001-d28-booster : 49 reads

Cutoff for CV-001-d7-booster : 146 reads

Cutoff for CV-001-d0-booster : 49 reads

Normalizing layer: counts

Normalizing across cells

Cutoff for CV-053-d28-vax : 20 reads

Cutoff for CV-053-d21-vax : 2 reads

Cutoff for CV-053-d7-vax : 8 reads

Cutoff for CV-053-d0-vax : 24 reads
</code></pre>
</div>
</div>
<p>Below we will visualize the output from HTODemux to inspect the quality of the demultiplexing algorithm. We quantified the percentage of “Singlet”, “Doublet” and “Negative” cells (empty droplets or cells with no tags) for each sample.</p>
<blockquote class="blockquote">
<p>HTODemux assigned &lt;50 % of cells as “Singlet”, which is not great.</p>
</blockquote>
<div id="cell-12" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># aggregate HTODemux output into a dataframe</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>metalist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="fu">seq_along</span>(obj.list)){</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    metalist[[x]] <span class="ot">&lt;-</span> obj.list[[x]]<span class="sc">@</span>meta.data}</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>metalist <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(metalist)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Quantify % of Singlet cells left from each sample</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>metalist_qc <span class="ot">&lt;-</span> metalist <span class="sc">%&gt;%</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples, HTO_classification.global) <span class="sc">%&gt;%</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">pct =</span> count<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">sum</span>(count)) <span class="sc">%&gt;%</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span>HTO_classification.global <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Doublet"</span>, <span class="st">"Negative"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">pct =</span> <span class="fu">sum</span>(pct))</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>metalist_qc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A tibble: 2 × 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">samples</th>
<th data-quarto-table-cell-role="th" scope="col">pct</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CV10</td>
<td>49.19421</td>
</tr>
<tr class="even">
<td>CV12</td>
<td>26.30427</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-13" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize % of singlet cells from each sample</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">5</span>, <span class="at">repr.plot.height=</span><span class="dv">4</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>metalist <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples, HTO_classification.global) <span class="sc">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">pct =</span> count<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">sum</span>(count)) <span class="sc">%&gt;%</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">y =</span> pct, <span class="at">fill =</span> HTO_classification.global)) <span class="sc">+</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_col</span>(<span class="at">color =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>HTO_classification.global) <span class="sc">+</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Proportions (%)"</span>) <span class="sc">+</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-8-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>In addition, we can inspect different parameter below to see if “Doublet” and “Negative” cells are assigned appropriately. Low nFeature/nCount typically indicate if the droplet is empty, hence “Negative”. Vice versa, high nFeature/nCount indicate “Doublet” cells as mRNA contents in a droplet is doubled.</p>
<blockquote class="blockquote">
<p>HTODemux assignments do reflect nature of the “Negative” and “Doublet” cells in our case looking at nFeature_RNA &amp; nCounts_RNA mainly.</p>
</blockquote>
<div id="cell-15" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect filtering parameters after HTODemux assignment</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">20</span>, <span class="at">repr.plot.height=</span> <span class="dv">6</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_HTO"</span>, <span class="st">"nCount_HTO"</span>, <span class="st">"pct.mt"</span>, <span class="st">"pct.rb"</span>, <span class="st">"pct.tcr"</span>, <span class="st">"pct.bcr"</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>metalist <span class="sc">%&gt;%</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(HTO_classification.global)) <span class="sc">%&gt;%</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(cols), <span class="at">names_to =</span> <span class="st">"measures"</span>, <span class="at">values_to =</span> <span class="st">"values"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">measures =</span> <span class="fu">factor</span>(measures, cols)) <span class="sc">%&gt;%</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> HTO_classification.global, <span class="at">y =</span> values, <span class="at">fill =</span> HTO_classification.global)) <span class="sc">+</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">bw =</span> <span class="st">"nrd0"</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="fl">0.75</span>), <span class="at">trim =</span> T, <span class="at">drop =</span> F, <span class="at">adjust =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples <span class="sc">+</span> measures, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-9-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-multiseqdemux" class="level2" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="sec-multiseqdemux"><span class="header-section-number">5.5</span> Run MULTIseqDemux</h2>
<p>Next we also will also implement MULTIseqDemux algorithm as a benchmark. HTODemux has been suggested to predict more false-positive doublets compared to MULTIseqDemux (see <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq/issues/26">GitHub</a> &amp; <a href="https://academic.oup.com/bioinformatics/article-abstract/38/10/2791/6565315">Paper</a>)</p>
<div id="cell-17" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run MULTIseqDemux</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(obj.list)){</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    obj.list[[x]] <span class="ot">&lt;-</span> <span class="fu">run_multiseqdemux</span>(obj.list[[x]], <span class="at">add.one =</span> F)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts

Normalizing across cells

Normalizing layer: counts

Normalizing across cells
</code></pre>
</div>
</div>
<p>Below we will visualize the output from MULTIseqDemux to inspect the quality of demultiplexing. Again we quantified the percentage of “Singlet”, “Doublet” and “Negative” cells for each sample.</p>
<blockquote class="blockquote">
<p>MUTLIseqDemux assigned &gt;80% and &gt;40 % of cells assigned as “Singlet” in CV10 and CV12 respectively, seemingly capturing more cells compared to HTODemux.</p>
</blockquote>
<div id="cell-19" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># aggregate MULTIseqDemux output into a dataframe</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>metalist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> <span class="fu">seq_along</span>(obj.list)){</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    metalist[[x]] <span class="ot">&lt;-</span> obj.list[[x]]<span class="sc">@</span>meta.data}</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>metalist <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(metalist)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># quantify % of Singlet cells left from each sample</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>metalist_qc <span class="ot">&lt;-</span> metalist <span class="sc">%&gt;%</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples, MULTI.global) <span class="sc">%&gt;%</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">pct =</span> count<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">sum</span>(count)) <span class="sc">%&gt;%</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span>MULTI.global <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Doublet"</span>, <span class="st">"Negative"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">pct =</span> <span class="fu">sum</span>(pct))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>metalist_qc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A tibble: 2 × 2</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th" scope="col">samples</th>
<th data-quarto-table-cell-role="th" scope="col">pct</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th" scope="col">&lt;chr&gt;</th>
<th data-quarto-table-cell-role="th" scope="col">&lt;dbl&gt;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CV10</td>
<td>84.33587</td>
</tr>
<tr class="even">
<td>CV12</td>
<td>42.67534</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-20" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize % of singlet cells left from each sample</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">5</span>, <span class="at">repr.plot.height=</span><span class="dv">4</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>metalist <span class="sc">%&gt;%</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples, MULTI.global) <span class="sc">%&gt;%</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarize</span>(<span class="at">count =</span> <span class="fu">n</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(samples) <span class="sc">%&gt;%</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">pct =</span> count<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">sum</span>(count)) <span class="sc">%&gt;%</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">y =</span> pct, <span class="at">fill =</span> MULTI.global)) <span class="sc">+</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>MULTI.global) <span class="sc">+</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Proportions (%)"</span>) <span class="sc">+</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Again, we can inspect different parameter below to see if “Doublet” and “Negative” cells are assigned appropriately.</p>
<blockquote class="blockquote">
<p>MULTIseqDemux assignments also reflect nature of the “Negative” and “Doublet” cells looking at nFeature_RNA &amp; nCounts_RNA mainly.</p>
</blockquote>
<div id="cell-22" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect filtering parameters after MULTIseqDemux assignment</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">20</span>, <span class="at">repr.plot.height=</span> <span class="dv">6</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_HTO"</span>, <span class="st">"nCount_HTO"</span>, <span class="st">"pct.mt"</span>, <span class="st">"pct.rb"</span>, <span class="st">"pct.tcr"</span>, <span class="st">"pct.bcr"</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>metalist <span class="sc">%&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(MULTI.global)) <span class="sc">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(cols), <span class="at">names_to =</span> <span class="st">"measures"</span>, <span class="at">values_to =</span> <span class="st">"values"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">measures =</span> <span class="fu">factor</span>(measures, cols)) <span class="sc">%&gt;%</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> MULTI.global, <span class="at">y =</span> values, <span class="at">fill =</span> MULTI.global)) <span class="sc">+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">bw =</span> <span class="st">"nrd0"</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="fl">0.75</span>), <span class="at">trim =</span> T, <span class="at">drop =</span> F, <span class="at">adjust =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples <span class="sc">+</span> measures, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-htosignal" class="level2" data-number="5.6">
<h2 data-number="5.6" class="anchored" data-anchor-id="sec-htosignal"><span class="header-section-number">5.6</span> Separate HTO Signals</h2>
<p>Furthermore we will inspect how well HTODemux and MULTIseqDemux separated the hashtag signals in both samples.</p>
<blockquote class="blockquote">
<p>MULTIseqDemux seems to perform slightly better in terms of separating the hashtag signals in both sequencing runs with minimal overlaps, especially for CV-053-d21-vax.</p>
</blockquote>
<div id="cell-24" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ridge plot of hashtag signals for CV10</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">20</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(obj.list[[<span class="dv">1</span>]], <span class="at">assay =</span> <span class="st">"HTO"</span>, <span class="at">group.by =</span> <span class="st">"hash.ID"</span>, <span class="at">features =</span> <span class="fu">rownames</span>(obj.list[[<span class="dv">1</span>]][[<span class="st">"HTO"</span>]]), <span class="at">ncol =</span> <span class="dv">4</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(obj.list[[<span class="dv">1</span>]], <span class="at">assay =</span> <span class="st">"HTO"</span>, <span class="at">group.by =</span> <span class="st">"MULTI_ID"</span>, <span class="at">features =</span> <span class="fu">rownames</span>(obj.list[[<span class="dv">1</span>]][[<span class="st">"HTO"</span>]]), <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0667

Picking joint bandwidth of 0.0643

Picking joint bandwidth of 0.109

Picking joint bandwidth of 0.0714

Picking joint bandwidth of 0.0729

Picking joint bandwidth of 0.0678

Picking joint bandwidth of 0.11

Picking joint bandwidth of 0.0698
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-14-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-14-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-25" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ridge plot of hashtag signals for CV12</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">20</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(obj.list[[<span class="dv">2</span>]], <span class="at">assay =</span> <span class="st">"HTO"</span>, <span class="at">group.by =</span> <span class="st">"hash.ID"</span>, <span class="at">features =</span> <span class="fu">rownames</span>(obj.list[[<span class="dv">2</span>]][[<span class="st">"HTO"</span>]]), <span class="at">ncol =</span> <span class="dv">4</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(obj.list[[<span class="dv">2</span>]], <span class="at">assay =</span> <span class="st">"HTO"</span>, <span class="at">group.by =</span> <span class="st">"MULTI_ID"</span>, <span class="at">features =</span> <span class="fu">rownames</span>(obj.list[[<span class="dv">2</span>]][[<span class="st">"HTO"</span>]]), <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0609

Picking joint bandwidth of 0.0685

Picking joint bandwidth of 0.0582

Picking joint bandwidth of 0.0799

Picking joint bandwidth of 0.0511

Picking joint bandwidth of 0.0622

Picking joint bandwidth of 0.044

Picking joint bandwidth of 0.0673
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-15-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.3_demux_files/figure-html/cell-15-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="evaluation" class="level2" data-number="5.7">
<h2 data-number="5.7" class="anchored" data-anchor-id="evaluation"><span class="header-section-number">5.7</span> Evaluation</h2>
<p>We will now merge the list of Seurat objects together as a single object to evaluate the demultiplexing algorithms.</p>
<blockquote class="blockquote">
<p>HTODemux and MULTIseqDemux algorithm outputs are mostly consistent. HTODemux called more negative cells than MULTIseqDemux.</p>
</blockquote>
<div id="cell-27" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge seurat objects and join all layers</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">merge</span>(obj.list[[<span class="dv">1</span>]], obj.list[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(obj.list)])</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">join_layers</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-28" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compare outputs from HTODemux and MULTIseqDemux</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">HTODemux=</span>obj<span class="sc">$</span>HTO_classification.global, <span class="at">MULTIseqDemux=</span>obj<span class="sc">$</span>MULTI.global)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>          MULTIseqDemux
HTODemux   Doublet Negative Singlet
  Doublet     4787      667    2043
  Negative     154     6089    8957
  Singlet      373      937   12843</code></pre>
</div>
</div>
<p>MULTIseqDemux seem to outperform HTODemux in terms of “Singlet” assignment and comparing hashtag signals. In the following chapters, we will continue subsequent analysis using MULTIseqDemux output rather than HTODemux.</p>
</section>
<section id="session-info" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="session-info">Session Info</h2>
<div id="cell-31" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(obj, <span class="st">"seurat/1_processing/1.3_GSE247917_demux.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-32" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Rocky Linux 8.7 (Green Obsidian)

Matrix products: default
BLAS/LAPACK: /nemo/lab/caladod/working/Matthew/.conda/envs/seurat5/lib/libopenblasp-r0.3.23.so;  LAPACK version 3.11.0

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/London
tzcode source: system (glibc)

attached base packages:
[1] stats4    grid      stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] scUnify_0.0.0.9000          ComplexHeatmap_2.16.0      
 [3] DoubletFinder_2.0.4         scDblFinder_1.14.0         
 [5] celda_1.16.1                Matrix_1.6-1               
 [7] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0
 [9] Biobase_2.62.0              GenomicRanges_1.54.1       
[11] GenomeInfoDb_1.38.0         IRanges_2.36.0             
[13] S4Vectors_0.40.1            BiocGenerics_0.48.0        
[15] MatrixGenerics_1.14.0       matrixStats_1.0.0          
[17] UCell_2.4.0                 BiocParallel_1.36.0        
[19] biomaRt_2.58.0              ggh4x_0.2.5                
[21] harmony_1.1.0               Rcpp_1.0.11                
[23] Seurat_5.0.0                SeuratObject_5.0.0         
[25] sp_2.1-1                    ggpubr_0.6.0               
[27] rstatix_0.7.2               writexl_1.4.2              
[29] readxl_1.4.3                qs_0.26.3                  
[31] RColorBrewer_1.1-3          wesanderson_0.3.6.9000     
[33] viridis_0.6.4               viridisLite_0.4.2          
[35] gridExtra_2.3               scales_1.3.0               
[37] cowplot_1.1.1               patchwork_1.2.0            
[39] ggrepel_0.9.4               lubridate_1.9.3            
[41] forcats_1.0.0               stringr_1.5.0              
[43] dplyr_1.1.4                 purrr_1.0.2                
[45] readr_2.1.4                 tidyr_1.3.0                
[47] tibble_3.2.1                ggplot2_3.5.1              
[49] tidyverse_2.0.0            

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.0-3      bitops_1.0-7              
  [3] httr_1.4.7                 doParallel_1.0.17         
  [5] repr_1.1.6                 tools_4.3.2               
  [7] sctransform_0.4.1          backports_1.4.1           
  [9] utf8_1.2.4                 R6_2.5.1                  
 [11] lazyeval_0.2.2             uwot_0.1.16               
 [13] GetoptLong_1.0.5           withr_3.0.1               
 [15] prettyunits_1.2.0          progressr_0.14.0          
 [17] cli_3.6.1                  spatstat.explore_3.2-5    
 [19] fastDummies_1.7.3          enrichR_3.2               
 [21] assertive.files_0.0-2      spatstat.data_3.0-3       
 [23] ggridges_0.5.4             pbapply_1.7-2             
 [25] Rsamtools_2.18.0           pbdZMQ_0.3-9              
 [27] scater_1.28.0              parallelly_1.36.0         
 [29] WriteXLS_6.4.0             limma_3.56.2              
 [31] RSQLite_2.3.2              BiocIO_1.12.0             
 [33] RApiSerialize_0.1.2        generics_0.1.3            
 [35] gridGraphics_0.5-1         shape_1.4.6               
 [37] combinat_0.0-8             ica_1.0-3                 
 [39] spatstat.random_3.2-1      car_3.1-2                 
 [41] ggbeeswarm_0.7.2           fansi_1.0.6               
 [43] abind_1.4-5                lifecycle_1.0.3           
 [45] edgeR_3.42.4               yaml_2.3.7                
 [47] carData_3.0-5              SparseArray_1.2.0         
 [49] BiocFileCache_2.10.1       Rtsne_0.16                
 [51] blob_1.2.4                 dqrng_0.3.1               
 [53] promises_1.2.1             crayon_1.5.2              
 [55] miniUI_0.1.1.1             lattice_0.21-8            
 [57] beachmat_2.18.0            KEGGREST_1.42.0           
 [59] magick_2.8.1               metapod_1.8.0             
 [61] pillar_1.9.0               rjson_0.2.21              
 [63] xgboost_1.7.7.1            future.apply_1.11.0       
 [65] codetools_0.2-19           leiden_0.4.3              
 [67] glue_1.8.0                 data.table_1.14.8         
 [69] vctrs_0.6.4                png_0.1-8                 
 [71] spam_2.10-0                cellranger_1.1.0          
 [73] gtable_0.3.4               cachem_1.0.8              
 [75] S4Arrays_1.2.0             mime_0.12                 
 [77] RcppEigen_0.3.3.9.3        survival_3.5-7            
 [79] iterators_1.0.14           statmod_1.5.0             
 [81] bluster_1.10.0             ellipsis_0.3.2            
 [83] fitdistrplus_1.1-11        ROCR_1.0-11               
 [85] nlme_3.1-163               bit64_4.0.5               
 [87] progress_1.2.2             filelock_1.0.2            
 [89] RcppAnnoy_0.0.21           irlba_2.3.5.1             
 [91] vipor_0.4.5                KernSmooth_2.23-22        
 [93] colorspace_2.1-0           DBI_1.1.3                 
 [95] tidyselect_1.2.0           bit_4.0.5                 
 [97] compiler_4.3.2             curl_5.1.0                
 [99] BiocNeighbors_1.18.0       assertive.properties_0.0-5
[101] xml2_1.3.5                 DelayedArray_0.28.0       
[103] plotly_4.10.3              rtracklayer_1.62.0        
[105] stringfish_0.16.0          lmtest_0.9-40             
[107] rappdirs_0.3.3             digest_0.6.33             
[109] goftest_1.2-3              spatstat.utils_3.0-5      
[111] XVector_0.42.0             htmltools_0.5.6.1         
[113] pkgconfig_2.0.3            assertive.numbers_0.0-2   
[115] base64enc_0.1-3            sparseMatrixStats_1.14.0  
[117] dbplyr_2.4.0               fastmap_1.1.1             
[119] rlang_1.1.4                GlobalOptions_0.1.2       
[121] htmlwidgets_1.6.2          DelayedMatrixStats_1.24.0 
[123] shiny_1.7.5.1              zoo_1.8-12                
[125] jsonlite_1.8.7             mclust_6.0.1              
[127] BiocSingular_1.16.0        RCurl_1.98-1.12           
[129] magrittr_2.0.3             scuttle_1.10.3            
[131] GenomeInfoDbData_1.2.11    dotCall64_1.1-0           
[133] multipanelfigure_2.1.2     IRkernel_1.3.2            
[135] munsell_0.5.0              reticulate_1.34.0         
[137] stringi_1.7.12             assertive.base_0.0-9      
[139] MCMCprecision_0.4.0        zlibbioc_1.48.0           
[141] MASS_7.3-60                plyr_1.8.9                
[143] parallel_4.3.2             listenv_0.9.0             
[145] deldir_1.0-9               Biostrings_2.70.1         
[147] IRdisplay_1.1              splines_4.3.2             
[149] tensor_1.5                 hms_1.1.3                 
[151] circlize_0.4.15            locfit_1.5-9.8            
[153] igraph_1.5.1               uuid_1.1-1                
[155] spatstat.geom_3.2-7        ggsignif_0.6.4            
[157] RcppHNSW_0.5.0             ScaledMatrix_1.8.1        
[159] reshape2_1.4.4             XML_3.99-0.14             
[161] evaluate_0.23              scran_1.28.2              
[163] RcppParallel_5.1.7         tzdb_0.4.0                
[165] foreach_1.5.2              httpuv_1.6.12             
[167] RANN_2.6.1                 polyclip_1.10-6           
[169] future_1.33.0              clue_0.3-64               
[171] scattermore_1.2            rsvd_1.0.5                
[173] broom_1.0.5                assertive.types_0.0-3     
[175] xtable_1.8-4               restfulr_0.0.15           
[177] RSpectra_0.16-1            dsb_1.0.4                 
[179] later_1.3.1                beeswarm_0.4.0            
[181] GenomicAlignments_1.38.0   memoise_2.0.1             
[183] AnnotationDbi_1.64.0       cluster_2.1.4             
[185] timechange_0.2.0           globals_0.16.2            </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../sections/1_processing/1.2_seurat.html" class="pagination-link  aria-label=" &lt;span="" object&lt;="" span&gt;"="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Seurat Object</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../sections/1_processing/1.4_qualitycontrol.html" class="pagination-link" aria-label="<span class='chapter-number'>6</span>&nbsp; <span class='chapter-title'>Quality Control</span>">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Quality Control</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>