<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Matthew Hung">
<meta name="dcterms.date" content="2024-11-07">

<title>Single-Cell Workbook - 7&nbsp; Doublet Detection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../sections/1_processing/1.4_qualitycontrol.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../sections/1_processing/1.1_cellranger.html"><strong>Chapter 1 : Data Processing</strong></a></li><li class="breadcrumb-item"><a href="../../sections/1_processing/1.5_doublets.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Doublet Detection</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-Cell Workbook</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/mshung229/scworkbook/">
            Report A Bug
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/mshung229">
            Follow Me
            </a>
          </li>
      </ul>
    </div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><b>Preface</b></span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text"><strong>Chapter 0 : Introduction</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/0_introduction/0.1_installation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Installation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/0_introduction/0.2_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Description</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text"><strong>Chapter 1 : Data Processing</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.1_cellranger.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">CellRanger (Multi)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.2_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Seurat Object</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.3_demux.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Cell Demultiplexing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.4_qualitycontrol.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/1_processing/1.5_doublets.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Doublet Detection</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text"><strong>Chapter 2 : Cell Classification</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../sections/2_cellclassification/2.1_sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">SCTransform</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#motivation" id="toc-motivation" class="nav-link active" data-scroll-target="#motivation">Motivation</a></li>
  <li><a href="#scdblfinder" id="toc-scdblfinder" class="nav-link" data-scroll-target="#scdblfinder"><span class="header-section-number">7.1</span> scDblFinder</a></li>
  <li><a href="#doubletfinder" id="toc-doubletfinder" class="nav-link" data-scroll-target="#doubletfinder"><span class="header-section-number">7.2</span> DoubletFinder</a></li>
  <li><a href="#tb-cell-doublets" id="toc-tb-cell-doublets" class="nav-link" data-scroll-target="#tb-cell-doublets"><span class="header-section-number">7.3</span> T/B-cell Doublets</a></li>
  <li><a href="#evaluating-outputs" id="toc-evaluating-outputs" class="nav-link" data-scroll-target="#evaluating-outputs"><span class="header-section-number">7.4</span> Evaluating Outputs</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../sections/1_processing/1.1_cellranger.html"><strong>Chapter 1 : Data Processing</strong></a></li><li class="breadcrumb-item"><a href="../../sections/1_processing/1.5_doublets.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Doublet Detection</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Doublet Detection</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Matthew Hung </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">2024-11-07</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="motivation" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="motivation">Motivation</h2>
<p>Similar to flow cytometry, a common issue with droplet-based single cell sequencing is the capture of multiple cells in a single droplet. The rate of doublet formation depends on the sequencing platform and the number of cells loaded. Below is a reference of doublet detection rate depending on no. of cells loaded during single-cell sequencing with the 10X chromium platform.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://kb.10xgenomics.com/hc/article_attachments/360086768291.png" class="img-fluid figure-img"></p>
<figcaption>10x Sequencing Doublet Detection Rate</figcaption>
</figure>
</div>
<p>Removing doublets is an important step in single-cell analysis to prevent false-positive findings. In the previous chapter, cell multiplexing with hashtag oligo is an effective method to identify and remove <b>homolytic and heterolytic</b> “Doublets”, however this requires sequencing of hashtag library which could be costly.</p>
<p>To overcome this issue there currently exists a few methods to identify “Doublets” based on mRNA expression, including DoubletFinder <a href="https://www.sciencedirect.com/science/article/pii/S2405471219300730?via%3Dihub">[McGinnis et al.; 2019]</a>, scDblFinder <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9204188/">[Germain et al, 2022]</a>, Scrublet <a href="https://www.sciencedirect.com/science/article/pii/S2405471218304745">[Wolock et al.; 2019]</a> etc. Many of these algorithms require users to process and cluster the cells prior to doublet prediction, as the identification process typically involves comparing the expressions from each cell to the pseudobulk expression profile of each cluster. These algorithms are effective in predicting <b>heterolytic</b> doublets, as detection works best when the pseudobulk profiles of the clusters are massively different. The pitfall, however, is that when you have a homogenous samples (i.e.&nbsp;FACS-sorted B-/Plasma cells) comprised of <b>homolytic</b> doublets, these detection algorithms may not generate accurate doublet predictions as the pseudobulk profiles of each cluster are similar.</p>
<blockquote class="blockquote">
<p>In our case, our data consist of innate and adaptive immune cells where gene expressions of the major lineages are very different. We will therefore predict doublets using DoubletFinder and scDblFinder to see if it provides addition benefit to remove potential heterolytic doublets. Refering to the table above, each of the samples have ~8k cells captured, therefore doublet rate is 6.4%.</p>
</blockquote>
<div id="cell-5" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## set up environment</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>({</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scUnify)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"/nemo/lab/caladod/working/Matthew/project/matthew/MH_GSE247917"</span>)})</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## load seurat objects</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">qread</span>(<span class="st">"seurat/1_processing/1.4_GSE247917_qc.qs"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(obj) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“replacing previous import ‘cowplot::get_legend’ by ‘ggpubr::get_legend’ when loading ‘scUnify’”
Warning message:
“replacing previous import ‘cowplot::align_plots’ by ‘patchwork::align_plots’ when loading ‘scUnify’”
Warning message:
“replacing previous import ‘biomaRt::select’ by ‘rstatix::select’ when loading ‘scUnify’”
Warning message:
“replacing previous import ‘scales::viridis_pal’ by ‘viridis::viridis_pal’ when loading ‘scUnify’”</code></pre>
</div>
</div>
</section>
<section id="scdblfinder" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="scdblfinder"><span class="header-section-number">7.1</span> scDblFinder</h2>
<p>We will first perform doublet detection with scDblFinder with the wrapper function below. The function will convert Seurat to SingleCellExperiment (SCE) object, run scDblFinder pipeline, and return a Seurat object with scDblFinder outputs stored in Seurat metadata. User can also specify cell demultiplexing results in the &lt;truth&gt; argument where the function will input “Singlet” and “Doublet” cells solely for doublet detection.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the case where only certain samples have cell multiplex libraries, users will need to set any “NA” values to “Singlets” for the wrapper function to run properly.</p>
</div>
</div>
<blockquote class="blockquote">
<p>In the run_scdblfinder() function, we specified cluster labels (“seurat_clusters”), sequencing runs (“samples”), doublet detection rate of 6.4% (dbr) and truth (“MULTI.global”).</p>
</blockquote>
<div id="cell-7" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>obj.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj, <span class="at">split.by =</span> <span class="st">"samples"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-8" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## convert to sce and run scDblFinder</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(obj.list)){</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    dbr <span class="ot">&lt;-</span> <span class="fu">calculate_10x_dbr</span>(<span class="fu">ncol</span>(obj.list[[i]]))</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    obj.list[[i]] <span class="ot">&lt;-</span> <span class="fu">run_scdblfinder</span>(</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>        obj.list[[i]], </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">clusters =</span> <span class="st">"integrated_snn_res.0.2"</span>, </span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">samples =</span> <span class="st">"samples"</span>, <span class="at">truth =</span> <span class="st">"MULTI.global"</span>, </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">dbr=</span>dbr, <span class="at">ncores =</span> <span class="dv">30</span>)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>returning assays (BCR, TCR) to RNA assay

Warning message:
“Different cells and/or features from existing assay RNA”
Warning message:
“Different cells and/or features from existing assay RNA”
Warning message:
“No layers found matching search pattern provided”
Warning message:
“No layers found matching search pattern provided”
Warning message:
“Layer ‘scale.data’ is empty”
Warning message:
“Assay RNA changing from Assay5 to Assay”
returning assays (BCR, TCR) to RNA assay

Warning message:
“Different cells and/or features from existing assay RNA”
Warning message:
“Different cells and/or features from existing assay RNA”
Warning message:
“No layers found matching search pattern provided”
Warning message:
“No layers found matching search pattern provided”
Warning message:
“Layer ‘scale.data’ is empty”
Warning message:
“Assay RNA changing from Assay5 to Assay”
R_zmq_msg_send errno: 4 strerror: Interrupted system call

</code></pre>
</div>
</div>
<p>Below we visualize the outputs of scDblFinder with MULTIseqDemux cell demultiplexing outputs and known characteristics of doublets (high nCount_RNA/HTO).</p>
<div id="cell-10" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">merge</span>(obj.list[[<span class="dv">1</span>]], obj.list[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(obj.list)], <span class="at">merge.dr =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging reduction ‘pca’
Merging reduction ‘umap’</code></pre>
</div>
</div>
<div id="cell-11" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">join_layers</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-12" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">## visualize scDblFinder output</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">18</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"scDblFinder"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"scDblFinder"</span>))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"MULTI.global"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"MULTIseqDemux"</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>plist[<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="fu">scFeaturePlot</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nCount_HTO"</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plist, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-7-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Compare to the cell demultiplexing outputs, scDblFinder detects and extra 926 doublets based on transcriptomic profile.</p>
</blockquote>
<div id="cell-14" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## compare cell multiplexing output (truth) to scDblFinder output (call)</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">truth=</span>obj<span class="sc">$</span>MULTI.global, <span class="at">call=</span>obj<span class="sc">$</span>scDblFinder)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>          call
truth      Doublet Singlet
  Doublet     1254    4060
  Negative       0       0
  Singlet     3297   20546</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>scDblFinder found 7.2% doublets (1139 cells) across the 2 sequencing runs.</p>
</blockquote>
<div id="cell-16" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## no. / percentage of doublets</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>no <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(obj<span class="sc">$</span>scDblFinder <span class="sc">==</span> <span class="st">"Doublet"</span>))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(no, <span class="st">" cells ("</span>, <span class="fu">round</span>(no<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">ncol</span>(obj), <span class="dv">1</span>), <span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
'4551 cells (12.4%)'
</div>
</div>
<p>Visualizing quality control parameters for singlets and doublets detected by scDblFinder.</p>
<blockquote class="blockquote">
<p>Doublets detected have higher nFeature_RNA &amp; nCount_RNA than singlets, reflecting the nature of true doublets.</p>
</blockquote>
<div id="cell-18" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Visualize qc metrics for doublets detected by scDblFinder</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">16</span>, <span class="at">repr.plot.height=</span> <span class="dv">6</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_HTO"</span>, <span class="st">"nCount_HTO"</span>, <span class="st">"pct.mt"</span>, <span class="st">"pct.rb"</span>, <span class="st">"pct.tcr"</span>, <span class="st">"pct.bcr"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(cols), <span class="at">names_to =</span> <span class="st">"measures"</span>, <span class="at">values_to =</span> <span class="st">"values"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">measures =</span> <span class="fu">factor</span>(measures, cols)) <span class="sc">%&gt;%</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(scDblFinder)) <span class="sc">%&gt;%</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> scDblFinder, <span class="at">y =</span> values, <span class="at">fill =</span> scDblFinder)) <span class="sc">+</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">bw =</span> <span class="st">"nrd0"</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="fl">0.75</span>), <span class="at">trim =</span> T, <span class="at">drop =</span> F, <span class="at">adjust =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples <span class="sc">+</span> measures, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.”</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-10-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="doubletfinder" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="doubletfinder"><span class="header-section-number">7.2</span> DoubletFinder</h2>
<p>Continuing from scDblFinder, we will split the merged Seurat object into a list of objects by sequencing runs in order to perform the DoubletFinder pipeline. In DoubletFinder, an optional argument available is take into consideration of the cell demultiplexing outputs during doublet detection and adjust for the no. of extra heterolytic doublets detected. This can be specified in the argument &lt;truth&gt; in the wrapper function below. The wrapper function will estimate an optimal pK value that is less than 0.1 (see <a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/62">https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/62</a>) to perform doublet detection.</p>
<blockquote class="blockquote">
<p>In the run_doubletfinder() function, we specified cluster labels (“seurat_clusters”), cell demultiplexing output (“MUTLI.global”), no. of PC used (1:10) and doublet detection rate of 6.4% (dbr).</p>
</blockquote>
<div id="cell-20" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## split seurat object by sequencing run and run doubletfinder</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">6</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>obj.list <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(obj, <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(obj.list)){</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    dbr <span class="ot">&lt;-</span> <span class="fu">calculate_10x_dbr</span>(<span class="fu">ncol</span>(obj.list[[i]]))</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    obj.list[[i]] <span class="ot">&lt;-</span> <span class="fu">run_doubletfinder</span>(</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>        obj.list[[i]], </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">assay =</span> <span class="st">"SCT"</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">truth =</span> <span class="st">"MULTI.global"</span>, </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">cluster =</span> <span class="st">"integrated_snn_res.0.2"</span>, </span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">dbr =</span> dbr, </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncores =</span> <span class="dv">30</span>)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Step 1 : Filter singlet and doublet from ground truth

Value of top pK is 0.001

Step 2 : Estimate Doublets

Loading required package: sctransform

Running SCTransform on assay: RNA

Running SCTransform on layer: counts

vst.flavor='v2' set. Using model with fixed slope and excluding poisson genes.

Variance stabilizing transformation of count matrix of size 22867 by 23432

Model formula is y ~ log_umi

Get Negative Binomial regression parameters per gene

Using 2000 genes, 5000 cells

Found 17 outliers - those will be ignored in fitting/regularization step


Second step: Get residuals using fitted parameters for 22867 genes

Computing corrected count matrix for 22867 genes

Calculating gene attributes

Wall clock passed: Time difference of 1.590437 mins

Determine variable features

Centering data matrix

Getting residuals for block 1(of 5) for counts dataset

Getting residuals for block 2(of 5) for counts dataset

Getting residuals for block 3(of 5) for counts dataset

Getting residuals for block 4(of 5) for counts dataset

Getting residuals for block 5(of 5) for counts dataset

Centering data matrix

Finished calculating residuals for counts

Set default assay to SCT

PC_ 1 
Positive:  LYZ, S100A9, CD74, FOS, VCAN, PSAP, IFI30, S100A8, FTL, FTH1 
       FCN1, CST3, CTSS, SAT1, NAMPT, NEAT1, SERPINA1, TYROBP, MNDA, CYBB 
       VIM, AIF1, THBS1, CD36, HLA-DRA, TNFAIP2, PLAUR, SPI1, ZFP36, DMXL2 
Negative:  IL7R, GNLY, NKG7, PRF1, MALAT1, LEF1, IL32, RPS12, CCL5, CD247 
       RPS3, CTSW, TCF7, ETS1, CD3E, CAMK4, IL2RG, RPL5, SKAP1, ITK 
       LTB, CD69, INPP4B, RPL10, SYNE2, RPL3, DDX5, RPS27A, NELL2, EVL 
PC_ 2 
Positive:  CD74, MS4A1, BANK1, JCHAIN, FCRL1, HLA-DRA, AFF3, RALGPS2, HLA-DQA1, CD79A 
       HLA-DPB1, HLA-DRB1, HLA-DPA1, LINC00926, HLA-DQB1, EEF2, FCRLA, PAX5, RPS8, EBF1 
       NIBAN3, WDFY4, ADAM28, BACH2, COL19A1, CDK14, BLK, RUBCNL, COBLL1, SWAP70 
Negative:  PPBP, GNLY, TUBB1, NKG7, PRF1, MPIG6B, F13A1, CCL5, CAVIN2, CTSW 
       ITGB3, CX3CR1, FCGR3A, GZMB, CD247, NRGN, THBS1, FGFBP2, MYOM2, FLNA 
       SPON2, CST7, IL2RB, ITGA2B, VCL, KLRD1, MYL9, SPARC, GZMA, SYNE1 
PC_ 3 
Positive:  GNLY, NKG7, PRF1, S100A9, LYZ, S100A8, CX3CR1, FOS, VCAN, FCGR3A 
       CTSW, ITGB2, CD247, GZMB, FGFBP2, AOAH, MYOM2, CST7, TYROBP, SPON2 
       PSAP, IL2RB, KLRD1, FTL, GZMA, S100A4, IFI30, SYNE1, FCN1, NEAT1 
Negative:  PPBP, TUBB1, MPIG6B, F13A1, CAVIN2, ITGB3, NRGN, ITGA2B, MYL9, SPARC 
       CD74, PTGS1, CLU, THBS1, RGS18, GP1BA, VCL, GNG11, GP9, MS4A1 
       LIMS1, BANK1, PDGFA, HIST1H2AC, STON2, MMRN1, JCHAIN, TAGLN2, GUCY1B1, PF4 
PC_ 4 
Positive:  CD74, JCHAIN, GNLY, PRF1, NKG7, MS4A1, BANK1, FCGR3A, CX3CR1, GZMB 
       FCRL1, CTSW, HLA-DRA, FGFBP2, MYOM2, CCL5, SPON2, PTGDS, HLA-DQA1, MZB1 
       HLA-DPA1, KLRD1, HLA-DPB1, CST7, IL2RB, CD79A, AFF3, RALGPS2, CD247, HLA-DRB1 
Negative:  S100A9, IL7R, LYZ, LEF1, AL138963.4, RPS12, TCF7, TPT1, S100A8, VCAN 
       CAMK4, NELL2, INPP4B, RPLP1, FOS, RPL13, RPL10, RPL32, RPS8, RPL4 
       CD3E, MAML2, RPS14, PDE3B, THEMIS, RPL30, RPL5, RPS3, RPS3A, PABPC1 
PC_ 5 
Positive:  CD74, MS4A1, BANK1, FCRL1, HLA-DRA, AFF3, HLA-DQA1, HLA-DPB1, HLA-DPA1, HLA-DRB1 
       RALGPS2, HLA-DQB1, GNLY, LINC00926, PRF1, NKG7, PAX5, FCRLA, COL19A1, WDFY4 
       ADAM28, SOX5, EBF1, HLA-DRB5, SWAP70, FCGR3A, CD22, NIBAN3, FCER2, CCSER1 
Negative:  JCHAIN, MZB1, HSP90B1, ITM2C, PPIB, S100A9, DERL3, PDIA4, SEC11C, TENT5C 
       IRF4, S100A8, KLHL14, LYZ, VCAN, PRDM1, HSPA5, RPLP0, XBP1, MKI67 
       FKBP11, FOS, AL589693.1, POU2AF1, CD38, FNDC3B, CD27, GAPDH, VIM, CALR 

Step 1 : Filter singlet and doublet from ground truth

Value of top pK is 0.001

Step 2 : Estimate Doublets

Running SCTransform on assay: RNA

Running SCTransform on layer: counts

vst.flavor='v2' set. Using model with fixed slope and excluding poisson genes.

Variance stabilizing transformation of count matrix of size 15106 by 15444

Model formula is y ~ log_umi

Get Negative Binomial regression parameters per gene

Using 950 genes, 5000 cells

Found 9 outliers - those will be ignored in fitting/regularization step


Second step: Get residuals using fitted parameters for 15106 genes

Computing corrected count matrix for 15106 genes

Calculating gene attributes

Wall clock passed: Time difference of 40.17004 secs

Determine variable features

Centering data matrix

Getting residuals for block 1(of 4) for counts dataset

Getting residuals for block 2(of 4) for counts dataset

Getting residuals for block 3(of 4) for counts dataset

Getting residuals for block 4(of 4) for counts dataset

Centering data matrix

Finished calculating residuals for counts

Set default assay to SCT

PC_ 1 
Positive:  HBB, TRIM58, SLC25A37, SLC25A39, ADIPOR1, UBB, FAM210B, YBX3, RNF10, MXI1 
       GSPT1, NCOA4, RBM38, MARCH8, GYPC, RIOK3, R3HDM4, BNIP3L, CDC34, FOXO3 
       FKBP8, UBE2H, PIP4K2A, PITHD1, RNF11, SIAH2, ARHGEF12, UBE2O, BSG, GUK1 
Negative:  RPL13, JUNB, COTL1, RPS8, MALAT1, RPL10, RPL12, RPL19, ZFP36L2, RPS2 
       RPL9, ACTB, HLA-B, RPL32, RPL34, HLA-E, EEF1A1, HLA-A, TMSB10, SPTAN1 
       AHNAK, RPL8, RPL13A, PFN1, PRRC2C, RPL5, H3F3B, RPL39, TMSB4X, TPT1 
PC_ 2 
Positive:  RPL13, LTB, TCF7, ZFP36L2, ETS1, MALAT1, TNFAIP3, RPL10, EEF1A1, RPS2 
       RPS27A, TLE5, RPL3, RPS12, RPS8, RPL19, TPT1, LEPROTL1, SUN2, RPL32 
       SARAF, IL32, RPS25, RPS18, KCNA3, RASGRP1, CD3E, RPS3A, NFATC2, RPS3 
Negative:  FTH1, CST3, COTL1, S100A9, CD74, CEBPD, LYZ, PSAP, ACTB, FCN1 
       FTL, MAFB, EGR1, TYROBP, IFI30, TYMP, SAT1, S100A4, DUSP1, LYN 
       MCL1, S100A6, S100A8, TNFAIP2, FOS, VIM, CTSZ, SPI1, CFD, VCAN 
PC_ 3 
Positive:  NKG7, GNLY, PRF1, EFHD2, TBX21, CTSW, SYNE1, CST7, FGFBP2, SPON2 
       RUNX3, FLNA, S1PR5, ITGB2, ARL4C, IGF2R, HLA-B, CCL5, HLA-C, ACTB 
       ABHD17A, PRKCH, SUN2, TGFBR3, TGFB1, GZMH, TPST2, MATK, HOPX, PFN1 
Negative:  RPL13, RPS2, RPS8, RPL10, LTB, TPT1, RPLP1, RPL19, RPL8, RPL32 
       RPL28, RPL12, RPS12, EEF1A1, FTH1, S100A9, RPL11, RPS27A, RPLP2, RPL18A 
       RPS18, RPL30, RPS15, RPL34, RPS9, RPS3A, RPS25, CST3, RPS23, RPL18 
PC_ 4 
Positive:  S100A9, S100A8, CEBPD, LYZ, MAFB, FP671120.4, MCL1, FCN1, FTH1, S100A4 
       VCAN, G0S2, FOS, EGR1, S100A6, TSPO, EFHD2, ATP2B1-AS1, TNFAIP3, GNLY 
       ZFP36L2, AC020916.1, NKG7, FLNA, VIM, FOSL2, ACSL1, TYROBP, IER3, HLA-A 
Negative:  CD74, CST3, BANK1, TNFRSF13C, TXNDC5, CCDC50, LYN, PTGDS, MT-CO1, JCHAIN 
       COTL1, REL, APP, PPP1R14B, EZR, BACH2, PSAP, SEL1L3, CHPT1, SAMHD1 
       HIST1H1E, PLCG2, ANKRD44, SYNGR2, PRKCB, CD37, FAM117B, CRYBG1, RNASET2, TRAK1 
PC_ 5 
Positive:  NKG7, GNLY, CST3, ACTB, TMSB4X, B2M, PFN1, PRF1, S100A4, CTSW 
       RPS2, CST7, EFHD2, RPL10, FGFBP2, TMSB10, ABHD17A, SH3BGRL3, CYBA, GZMM 
       RPS19, TBX21, CD81, SPON2, CCL5, HOPX, RPL13, CALM1, IFITM1, CFL1 
Negative:  MCL1, TCF7, MALAT1, EGR1, TPT1, PABPC1, AC020916.1, S100A9, FOS, VCAN 
       LYZ, JUNB, SMCHD1, PIK3IP1, ACSL1, NFKBIZ, DUSP1, LEF1, TNFAIP2, RCAN3 
       TSC22D3, PDE3B, HOOK2, TNFAIP3, ATP2B1, PLK3, CDKN1B, MAML2, DDX5, BACH2 
</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>NULL
[1] "Creating 5858 artificial doublets..."
[1] "Creating Seurat object..."
[1] "Running SCTransform..."
[1] "Running PCA..."
[1] "Calculating PC distance matrix..."
[1] "Computing pANN..."
[1] "Classifying doublets.."
NULL
[1] "Creating 3861 artificial doublets..."
[1] "Creating Seurat object..."
[1] "Running SCTransform..."
[1] "Running PCA..."
[1] "Calculating PC distance matrix..."
[1] "Computing pANN..."
[1] "Classifying doublets.."</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-11-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-11-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-21" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## merge seurat objects back to a single seurat object</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">merge</span>(obj.list[[<span class="dv">1</span>]], obj.list[<span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(obj.list)], <span class="at">merge.dr =</span> T)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>obj <span class="ot">&lt;-</span> <span class="fu">join_layers</span>(obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging reduction ‘pca’
Merging reduction ‘umap’</code></pre>
</div>
</div>
<p>Below we visualize the outputs of DoubletFinder with MULTIseqDemux cell demultiplexing outputs and known characteristics of doublets (high nCount_RNA/HTO).</p>
<div id="cell-23" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Visualize doubletfinder output with multiseq demux and nCount_RNA/HTO</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">18</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"DoubletFinder"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"DoubletFinder"</span>))</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"MULTI.global"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"MULTIseqDemux"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>plist[<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="fu">scFeaturePlot</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nCount_HTO"</span>), <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plist, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Compare to the cell demultiplexing outputs, DoubletFinder detects and extra 745 doublets based on transcriptomic profile.</p>
</blockquote>
<div id="cell-25" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compare cell multiplexing output (truth) to doubletfinder output (call)</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">truth=</span>obj<span class="sc">$</span>MULTI.global, <span class="at">call=</span>obj<span class="sc">$</span>DoubletFinder)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>          call
truth      Doublet Singlet
  Doublet     1162    4152
  Negative       0       0
  Singlet     3286   20557</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>DoubletFinder found 5.6% doublets (880 cells) across the 2 sequencing runs.</p>
</blockquote>
<div id="cell-27" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="21">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## no. / percentage of doublets</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>no <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(obj<span class="sc">$</span>DoubletFinder <span class="sc">==</span> <span class="st">"Doublet"</span>))</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paste0</span>(no, <span class="st">" cells ("</span>, <span class="fu">round</span>(no<span class="sc">*</span><span class="dv">100</span><span class="sc">/</span><span class="fu">ncol</span>(obj), <span class="dv">1</span>), <span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
'4448 cells (12.1%)'
</div>
</div>
<p>Visualizing quality control parameters for singlets and doublets detected by DoubletFinder.</p>
<blockquote class="blockquote">
<p>Similar to scDblFinder, doublets detected in DoubletFinder have higher nFeature_RNA &amp; nCount_RNA than singlets, reflecting the nature of true doublets.</p>
</blockquote>
<div id="cell-29" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="17">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Visualize qc metrics for doublets detected by DoubletFinder</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">16</span>, <span class="at">repr.plot.height=</span> <span class="dv">6</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_HTO"</span>, <span class="st">"nCount_HTO"</span>, <span class="st">"pct.mt"</span>, <span class="st">"pct.rb"</span>, <span class="st">"pct.tcr"</span>, <span class="st">"pct.bcr"</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(DoubletFinder)) <span class="sc">%&gt;%</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(cols), <span class="at">names_to =</span> <span class="st">"measures"</span>, <span class="at">values_to =</span> <span class="st">"values"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">measures =</span> <span class="fu">factor</span>(measures, cols)) <span class="sc">%&gt;%</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> DoubletFinder, <span class="at">y =</span> values, <span class="at">fill =</span> DoubletFinder)) <span class="sc">+</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">bw =</span> <span class="st">"nrd0"</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="fl">0.75</span>), <span class="at">trim =</span> T, <span class="at">drop =</span> F, <span class="at">adjust =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples <span class="sc">+</span> measures, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-16-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="tb-cell-doublets" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="tb-cell-doublets"><span class="header-section-number">7.3</span> T/B-cell Doublets</h2>
<p>T/B-cells represent the most abundant of immune cell populations and have a high chance to be sequenced together as doublets. T and B-cells can be characterized by TCR-VDJ and BCR-VDJ gene expression respectively, the TCR/BCR expressions are mutually exclusive. Therefore T/B-cell doublets can be manually identified and removed if the TCR/BCR fraction is high.</p>
<div id="cell-31" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="18">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="do">## visualize BCR/TCR read fractions</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">9</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> pct.tcr, <span class="at">y =</span> pct.bcr)) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.05</span>) <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"TCR Fraction"</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"BCR Fraction"</span>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(pct.tcr), <span class="at">y =</span> <span class="fu">log10</span>(pct.bcr))) <span class="sc">+</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fu">log10</span>(<span class="dv">5</span>)) <span class="sc">+</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">log10</span>(<span class="fl">0.05</span>)) <span class="sc">+</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">"log10(TCR Fraction)"</span>) <span class="sc">+</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"log10(BCR Fraction)"</span>)</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-17-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p>Judging from the plot above, we have called T/B-cell doublets if TCR fraction is &gt; 0.05% and BCR &gt; 5%</p>
</blockquote>
<div id="cell-33" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="19">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="do">## manually set TCR/BCR fraction threshold</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">TB_Doublet =</span> <span class="fu">ifelse</span>(pct.tcr <span class="sc">&gt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> pct.bcr <span class="sc">&gt;</span> <span class="dv">5</span>, <span class="st">"Doublet"</span>, <span class="st">"Singlet"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-34" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="20">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Visualize qc metrics for TB doublets</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">16</span>, <span class="at">repr.plot.height=</span> <span class="dv">6</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_HTO"</span>, <span class="st">"nCount_HTO"</span>, <span class="st">"pct.mt"</span>, <span class="st">"pct.rb"</span>, <span class="st">"pct.tcr"</span>, <span class="st">"pct.bcr"</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>obj<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(cols), <span class="at">names_to =</span> <span class="st">"measures"</span>, <span class="at">values_to =</span> <span class="st">"values"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">measures =</span> <span class="fu">factor</span>(measures, cols)) <span class="sc">%&gt;%</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> TB_Doublet, <span class="at">y =</span> values, <span class="at">fill =</span> TB_Doublet)) <span class="sc">+</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="at">size =</span> <span class="dv">1</span>, <span class="at">scale =</span> <span class="st">"width"</span>, <span class="at">bw =</span> <span class="st">"nrd0"</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="fl">0.75</span>), <span class="at">trim =</span> T, <span class="at">drop =</span> F, <span class="at">adjust =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples <span class="sc">+</span> measures, <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">ncol =</span> <span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">""</span>)) <span class="sc">+</span></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_border</span>() <span class="sc">+</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_aes</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="evaluating-outputs" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="evaluating-outputs"><span class="header-section-number">7.4</span> Evaluating Outputs</h2>
<p>scDblFinder and DoubletFinder are both able to predict heterolytic doublets with characteristics reflective of true doublet nature. From the UMAP projections below, you can see the outputs from the 2 algorithms are highly consistent with each other.</p>
<blockquote class="blockquote">
<p>The small cluster of cells circled on the UMAP below are highly likely to be real doublets given the cell number, which is only detected by DoubletFinder and not scDblFinder. Therefore, for the following chapters, we will continue using outputs from DoubletFinder.</p>
</blockquote>
<div id="cell-36" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="25">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">18</span>, <span class="at">repr.plot.height=</span> <span class="dv">4</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>r <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>xc <span class="ot">=</span> <span class="fl">5.5</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>yc <span class="ot">=</span> <span class="fl">7.5</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"DoubletFinder"</span>) <span class="sc">+</span> </span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"DoubletFinder"</span>) <span class="co">#+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#annotate("path",</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    x=xc+r*cos(seq(0,2*pi,length.out=100)),</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    y=yc+r*sin(seq(0,2*pi,length.out=100)),</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    linetype = "dashed")</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"scDblFinder"</span>) <span class="sc">+</span> </span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"scDblFinder"</span>) <span class="co">#+</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">#annotate("path",</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    x=xc+r*cos(seq(0,2*pi,length.out=100)),</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    y=yc+r*sin(seq(0,2*pi,length.out=100)),</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    linetype = "dashed")</span></span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"TB_Doublet"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"T/B Doublet"</span>)</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>plist[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> <span class="fu">scUMAP</span>(obj, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"MULTI.global"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"MULTIseqDemux"</span>)</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plist, <span class="at">ncol =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="1.5_doublets_files/figure-html/cell-20-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>~1/3 of the doublets detected by either algorithms are consistent.</p>
<div id="cell-38" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="22">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Compare outputs from scDblFinder and DoubletFinder</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">scDblFinder=</span>obj<span class="sc">$</span>scDblFinder, <span class="at">DoubletFinder=</span>obj<span class="sc">$</span>DoubletFinder)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>           DoubletFinder
scDblFinder Doublet Singlet
    Doublet    2355    2196
    Singlet    2093   22513</code></pre>
</div>
</div>
</section>
<section id="session-info" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="session-info">Session Info</h2>
<div id="cell-40" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="26">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qsave</span>(obj, <span class="at">file =</span> <span class="st">"seurat/1_processing/1.5_GSE247917_doublets.qs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-41" class="cell" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Rocky Linux 8.7 (Green Obsidian)

Matrix products: default
BLAS/LAPACK: /nemo/lab/caladod/working/Matthew/.conda/envs/seurat5/lib/libopenblasp-r0.3.23.so;  LAPACK version 3.11.0

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/London
tzcode source: system (glibc)

attached base packages:
 [1] parallel  stats4    grid      stats     graphics  grDevices utils    
 [8] datasets  methods   base     

other attached packages:
 [1] sctransform_0.4.1           ROCR_1.0-11                
 [3] KernSmooth_2.23-22          fields_15.2                
 [5] spam_2.10-0                 scUnify_0.0.0.9000         
 [7] ComplexHeatmap_2.16.0       DoubletFinder_2.0.4        
 [9] scDblFinder_1.14.0          celda_1.16.1               
[11] Matrix_1.6-1                SingleCellExperiment_1.24.0
[13] SummarizedExperiment_1.32.0 Biobase_2.62.0             
[15] GenomicRanges_1.54.1        GenomeInfoDb_1.38.0        
[17] IRanges_2.36.0              S4Vectors_0.40.1           
[19] BiocGenerics_0.48.0         MatrixGenerics_1.14.0      
[21] matrixStats_1.0.0           UCell_2.4.0                
[23] BiocParallel_1.36.0         biomaRt_2.58.0             
[25] ggh4x_0.2.5                 harmony_1.1.0              
[27] Rcpp_1.0.11                 Seurat_5.0.0               
[29] SeuratObject_5.0.0          sp_2.1-1                   
[31] ggpubr_0.6.0                rstatix_0.7.2              
[33] writexl_1.4.2               readxl_1.4.3               
[35] qs_0.26.3                   RColorBrewer_1.1-3         
[37] wesanderson_0.3.6.9000      viridis_0.6.4              
[39] viridisLite_0.4.2           gridExtra_2.3              
[41] scales_1.3.0                cowplot_1.1.1              
[43] patchwork_1.2.0             ggrepel_0.9.4              
[45] lubridate_1.9.3             forcats_1.0.0              
[47] stringr_1.5.0               dplyr_1.1.4                
[49] purrr_1.0.2                 readr_2.1.4                
[51] tidyr_1.3.0                 tibble_3.2.1               
[53] ggplot2_3.5.1               tidyverse_2.0.0            

loaded via a namespace (and not attached):
  [1] progress_1.2.2             goftest_1.2-3             
  [3] Biostrings_2.70.1          vctrs_0.6.4               
  [5] spatstat.random_3.2-1      RApiSerialize_0.1.2       
  [7] digest_0.6.33              png_0.1-8                 
  [9] shape_1.4.6                IRdisplay_1.1             
 [11] deldir_1.0-9               parallelly_1.36.0         
 [13] combinat_0.0-8             magick_2.8.1              
 [15] MASS_7.3-60                reshape2_1.4.4            
 [17] httpuv_1.6.12              foreach_1.5.2             
 [19] withr_3.0.1                ellipsis_0.3.2            
 [21] survival_3.5-7             memoise_2.0.1             
 [23] ggbeeswarm_0.7.2           zoo_1.8-12                
 [25] GlobalOptions_0.1.2        pbapply_1.7-2             
 [27] IRkernel_1.3.2             prettyunits_1.2.0         
 [29] KEGGREST_1.42.0            promises_1.2.1            
 [31] httr_1.4.7                 restfulr_0.0.15           
 [33] globals_0.16.2             fitdistrplus_1.1-11       
 [35] stringfish_0.16.0          dsb_1.0.4                 
 [37] miniUI_0.1.1.1             generics_0.1.3            
 [39] base64enc_0.1-3            curl_5.1.0                
 [41] repr_1.1.6                 zlibbioc_1.48.0           
 [43] ScaledMatrix_1.8.1         polyclip_1.10-6           
 [45] GenomeInfoDbData_1.2.11    SparseArray_1.2.0         
 [47] RcppEigen_0.3.3.9.3        xtable_1.8-4              
 [49] doParallel_1.0.17          evaluate_0.23             
 [51] S4Arrays_1.2.0             BiocFileCache_2.10.1      
 [53] hms_1.1.3                  irlba_2.3.5.1             
 [55] colorspace_2.1-0           filelock_1.0.2            
 [57] reticulate_1.34.0          spatstat.data_3.0-3       
 [59] magrittr_2.0.3             lmtest_0.9-40             
 [61] glmGamPoi_1.14.0           later_1.3.1               
 [63] lattice_0.21-8             spatstat.geom_3.2-7       
 [65] future.apply_1.11.0        scattermore_1.2           
 [67] XML_3.99-0.14              scuttle_1.10.3            
 [69] assertive.numbers_0.0-2    RcppAnnoy_0.0.21          
 [71] pillar_1.9.0               nlme_3.1-163              
 [73] iterators_1.0.14           compiler_4.3.2            
 [75] beachmat_2.18.0            RSpectra_0.16-1           
 [77] stringi_1.7.12             assertive.properties_0.0-5
 [79] tensor_1.5                 GenomicAlignments_1.38.0  
 [81] MCMCprecision_0.4.0        plyr_1.8.9                
 [83] crayon_1.5.2               abind_1.4-5               
 [85] BiocIO_1.12.0              scater_1.28.0             
 [87] gridGraphics_0.5-1         locfit_1.5-9.8            
 [89] bit_4.0.5                  codetools_0.2-19          
 [91] BiocSingular_1.16.0        GetoptLong_1.0.5          
 [93] plotly_4.10.3              mime_0.12                 
 [95] splines_4.3.2              circlize_0.4.15           
 [97] fastDummies_1.7.3          dbplyr_2.4.0              
 [99] sparseMatrixStats_1.14.0   cellranger_1.1.0          
[101] blob_1.2.4                 utf8_1.2.4                
[103] clue_0.3-64                pbdZMQ_0.3-9              
[105] WriteXLS_6.4.0             listenv_0.9.0             
[107] DelayedMatrixStats_1.24.0  ggsignif_0.6.4            
[109] assertive.base_0.0-9       statmod_1.5.0             
[111] tzdb_0.4.0                 pkgconfig_2.0.3           
[113] tools_4.3.2                cachem_1.0.8              
[115] RSQLite_2.3.2              DBI_1.1.3                 
[117] fastmap_1.1.1              ica_1.0-3                 
[119] Rsamtools_2.18.0           broom_1.0.5               
[121] dotCall64_1.1-0            carData_3.0-5             
[123] RANN_2.6.1                 farver_2.1.1              
[125] yaml_2.3.7                 rtracklayer_1.62.0        
[127] cli_3.6.1                  assertive.types_0.0-3     
[129] leiden_0.4.3               lifecycle_1.0.3           
[131] uwot_0.1.16                bluster_1.10.0            
[133] assertive.files_0.0-2      backports_1.4.1           
[135] timechange_0.2.0           gtable_0.3.4              
[137] rjson_0.2.21               ggridges_0.5.4            
[139] progressr_0.14.0           limma_3.56.2              
[141] jsonlite_1.8.7             edgeR_3.42.4              
[143] RcppHNSW_0.5.0             bitops_1.0-7              
[145] bit64_4.0.5                xgboost_1.7.7.1           
[147] Rtsne_0.16                 spatstat.utils_3.0-5      
[149] BiocNeighbors_1.18.0       RcppParallel_5.1.7        
[151] metapod_1.8.0              dqrng_0.3.1               
[153] enrichR_3.2                lazyeval_0.2.2            
[155] shiny_1.7.5.1              htmltools_0.5.6.1         
[157] rappdirs_0.3.3             glue_1.8.0                
[159] XVector_0.42.0             RCurl_1.98-1.12           
[161] scran_1.28.2               mclust_6.0.1              
[163] igraph_1.5.1               R6_2.5.1                  
[165] labeling_0.4.3             cluster_2.1.4             
[167] DelayedArray_0.28.0        tidyselect_1.2.0          
[169] vipor_0.4.5                maps_3.4.2                
[171] xml2_1.3.5                 car_3.1-2                 
[173] AnnotationDbi_1.64.0       future_1.33.0             
[175] rsvd_1.0.5                 munsell_0.5.0             
[177] multipanelfigure_2.1.2     data.table_1.14.8         
[179] htmlwidgets_1.6.2          rlang_1.1.4               
[181] spatstat.sparse_3.0-3      spatstat.explore_3.2-5    
[183] uuid_1.1-1                 fansi_1.0.6               
[185] beeswarm_0.4.0            </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../sections/1_processing/1.4_qualitycontrol.html" class="pagination-link  aria-label=" &lt;span="" control&lt;="" span&gt;"="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Quality Control</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->




</body></html>